pkg/local_object_storage/writecache/flush.go

```diff
@@ -99,10 +99,6 @@ func (c *cache) flushDB() {
 					lastKey = slice.Copy(k)
 				}
 
-				if _, ok := c.flushed.Peek(string(k)); ok {
```

Comment:
Why don't you just use `Contains()` instead?

Reply:
Also, `Peek` is not supposed to delete anything.

Reply:
I didn't knew about `Contains`, fixed!

`Peek` doesn't delete but it takes read mutex _inside_ of a `View` transaction. `View` transaction itself takes `mmapLock`: https://github.com/etcd-io/bbolt/blob/8f4a7e1f92975fbd1536a569d7aad6800809ef4e/db.go#L708
which is lifted after tx finishes (in `tx.Commit()` -> `tx.close()` -> `tx.db.removeTx`)

When we evict items from LRU cache mutex order is different: first we lock cache and then execute `Batch` which usually doesn't take `mmapLock`. However it case we need to remap it _does_. Thus the deadlock.

Added this to a commit message.

Reply:
IIUC, `flushed` is a _simple_ LRU as in https://github.com/hashicorp/golang-lru/blob/v0.6.0/simplelru/lru.go#L93. Where does it take any read mutex?

You're writing "eviction handler removes untracked items from the database". But neither Peek nor Contains remove items from the LRU cache. Am I missing something?

Reply:
`simplelru` package is `LRU` (simple LRU) and `LRUCache` (an interface). We use the latter.
https://github.com/hashicorp/golang-lru/blob/v0.6.0/simplelru/lru_interface.go

Yes, but we _can_ execute `flushed.Add` in another goroutine in which case it _will_ evict.